src/compiler/scala/tools/nsc/transform/TypeAdaptingTransformer.scala

```diff
@@ -88,17 +88,17 @@ trait TypeAdaptingTransformer { self: TreeDSL =>
           if (treeInfo isExprSafeToInline side) value
           else BLOCK(side, value)
         val tree1 = pt match {
-          case ErasedValueType(clazz, BoxedUnitTpe) =>
-            cast(preservingSideEffects(tree, REF(BoxedUnit_UNIT)), pt)
-          case ErasedValueType(clazz, underlying) => cast(unboxValueClass(tree, clazz, underlying), pt)
+          case ErasedValueType(clazz, BoxedUnitTpe) => cast(preservingSideEffects(tree, REF(BoxedUnit_UNIT)), pt)
+          case ErasedValueType(clazz, underlying)   => cast(unboxValueClass(tree, clazz, underlying), pt)
           case _ =>
             pt.typeSymbol match {
               case UnitClass  =>
                 preservingSideEffects(tree, UNIT)
-              case x          =>
-                assert(x != ArrayClass, "array")
-                // don't `setType pt` the Apply tree, as the Apply's fun won't be typechecked if the Apply tree already has a type
-                Apply(currentRun.runDefinitions.unboxMethod(pt.typeSymbol), tree)
+              case ArrayClass => assert(pt.typeSymbol != ArrayClass, "array") ; tree
+              case _          =>
+                val unboxer = currentRun.runDefinitions.unboxMethod(pt.typeSymbol)
```

Comment:
This, I assume fixes https://github.com/scala/bug/issues/10539#issuecomment-526781414, right?

Did you mean to add that case?  Would be nice to have the case that accompanies this change, I think.

Reply:
Without the fix, under -Xdev the test says `  String is not a boxed Boolean`. The case with a primitive result such as Int still fails in backend with `  inconvertible types : I -> Z`

Reply:
BTW the only change here is the assert under -Xdev. Otherwise it's just syntax for ease of reading.

Reply:
Oh, yeah... 🤔 Not sure why I thought this was introducing more unboxing.